pkg/kubectl/cmd/drain.go

```diff
@@ -0,0 +1,159 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/meta"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/fields"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/labels"
+)
+
+type DrainOptions struct {
+	NodeName           string
+	Force              bool
+	GracePeriodSeconds int
+}
+
+const (
+	drain_long = `Drain nodes in preparation for maintenance.
+
+`
+	drain_example = `# Drain node "foo", even if there are unreplicated pods on it.
+$ kubectl drain foo --force
+
+# As above, but don't delete unreplicated pods, and use a grace period of 15 minutes.
+$ kubectl drain foo --grace-period=900
+`
+)
+
+func NewCmdDrain(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &DrainOptions{}
+
+	cmd := &cobra.Command{
+		Use:     "drain NODE",
+		Short:   "Drain nodes in preparation for maintenance",
+		Long:    drain_long,
+		Example: drain_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(
+				RunDrain(f, out, cmd, args, options))
+		},
+	}
+	cmd.Flags().BoolVar(&options.Force, "force", false, "Delete unreplicated pods.")
+	cmd.Flags().IntVar(&options.GracePeriodSeconds, "grace-period", -1, "Period of time in seconds given to each pod to terminate gracefully. Ignored if negative.")
+	return cmd
+}
+
+func RunDrain(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string, options *DrainOptions) error {
+	if len(args) != 1 {
+		return cmdutil.UsageError(cmd, fmt.Sprintf("USAGE: %s [flags]", cmd.Use))
+	}
+	options.NodeName = args[0]
+
+	mapper, _ := f.Object()
+	c, err := f.Client()
+	if err != nil {
+		return err
+	}
+	if err = maybeLameNode(mapper, c, out, options); err != nil {
+		return err
+	}
+	pods, err := c.Pods(api.NamespaceAll).List(nil, fields.Set{"spec.nodeName": options.NodeName}.AsSelector())
+	if err != nil {
+		return err
+	}
+	if err = scanForReplicationControllers(c, out, options, pods); err != nil {
+		return err
+	}
+	if err = deletePods(mapper, c, out, options, pods); err != nil {
+		return err
+	}
+	cmdutil.PrintSuccess(mapper, false, out, "node", options.NodeName, "drained")
+	return nil
+}
+
+func maybeLameNode(mapper meta.RESTMapper, c *client.Client, out io.Writer, options *DrainOptions) error {
+	node, err := c.Nodes().Get(options.NodeName)
```

Comment:
On Wed, Nov 4, 2015 at 9:17 PM, Brian Grant <notifications@github.com>
wrote:

> In pkg/kubectl/cmd/drain.go
> <https://github.com/kubernetes/kubernetes/pull/16698#discussion_r43977179>
> :
>
> > +	pods, err := c.Pods(api.NamespaceAll).List(nil, fields.Set{"spec.nodeName": options.NodeName}.AsSelector())
> > +	if err != nil {
> > +		return err
> > +	}
> > +	if err = scanForReplicationControllers(c, out, options, pods); err != nil {
> > +		return err
> > +	}
> > +	if err = deletePods(mapper, c, out, options, pods); err != nil {
> > +		return err
> > +	}
> > +	cmdutil.PrintSuccess(mapper, false, out, "node", options.NodeName, "drained")
> > +	return nil
> > +}
> > +
> > +func maybeLameNode(mapper meta.RESTMapper, c *client.Client, out io.Writer, options *DrainOptions) error {
> > +	node, err := c.Nodes().Get(options.NodeName)
>
> We should at least support multiple nodes and selection of nodes via label
> selector.
>

I was intentionally dodging that because it forces us to expose semantics
for what happens in the case of errors.  (Mentioned in the original issue.)
 E.g.: Do we search for unreplicated pods on all nodes before starting?  In
the case of a failure or partial failure, do we continue with the other
nodes?

Regarding selecting via selector, once I modify it for node/foo usage, the
user can do the below.  I consider this kind of composability good and
would like to see documented examples written in this style.

Inevitably we'll need to address multiple machines, but for now I'm
focusing on getting the simplest thing shipped that will make things better.


for node in $(kubectl get nodes -o name -l kubernetes.io/foo=bar); do

kubectl drain $node --grace-period=120

sleep 900

done


—
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16698/files#r43977179>.
>